### ISSUE TITLE
[Ci] Choose a fixed seed for Private CI regression runs

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#1735]
Private CI runs regressions with a random seed, which makes tests flaky (the goal of the regression run is to give results that depend only on the change being tested, and remove most other factors). 

@sriyerg Is there any specific seed we should be using? Or can we just go with `1` or something like that?

**Comment by gkelly** [comment]
On software tests we'd traditionally use a random seed, because if a test is flaky then it's indicative of some other issue: bad test, actual bad behavior for a given seed, etc.  Is the fixed seed in this case just to prevent test timeouts?

**Comment by imphil** [comment]
Test timeouts (or more general: very different runtime depending on the seed) is one concern, but not the main one here.

The main motivation for fixing the seed is a separation of concerns: 

- The sanity run, as we run it in CI, is meant to answer the question "is the tested code better or worse than the previous one". To answer that, we need to apply the same "rules", i.e. tests. 
- To perform a more in-depth exploration of the design at a single point we need to randomize. That aspect is covered in the nightly (and full) regression runs.

Constraint random testing as we do for DV adds significantly more unpredictability to the test runs than most software testing techniques do (fuzzing being a prominent exception to that), which makes "fixing the seed" more important.

In more practical terms, it is not a great commit experience for a software engineer to file a bug for a hardware bug, which happens to show up with a given seed, before being able to commit a software change.

**Comment by sriyerg** [comment]
Agree that we should be fixing the seed. I will make an update in the tool to add support for fixing the seed across all tests.

**Comment by sriyerg** [comment]
I also agree with @gkelly that flakiness (tests passing in some seeds and failing in other) is indicative of a real issue (either in the design or DV) that definitely needs to be addressed. It is for that reason that we run each tests in the nightly regressions with 100s of seeds. Our DV methodology mandates 100% passing on all seeds with a 7-day soak time before we can sign-off on a particular block (hence the 7 day result history). 

Since we are already running with a lot of randomization in our nightly regressions, we can relax the CI and make it more deterministic, not to avoid flakiness, but mainly to make the run times consistent and within a reasonable timeout (currently set to 15 minutes). Our main goal is to only ensure we did not break anything, not to actually find bugs. 

We typically choose the fixed seed value as 1 for sanity CI, but the design and DV are in constant flux. We may have to adjust the tests (or adjust the seed even) depending on what the run time looks like.

**Comment by gkelly** [comment]
> I also agree with @gkelly that flakiness (tests passing in some seeds and failing in other) is indicative of a real issue (either in the design or DV) that definitely needs to be addressed. It is for that reason that we run each tests in the nightly regressions with 100s of seeds. Our DV methodology mandates 100% passing on all seeds with a 7-day soak time before we can sign-off on a particular block (hence the 7 day result history).
> 
> Since we are already running with a lot of randomization in our nightly regressions, we can relax the CI and make it more deterministic, not to avoid flakiness, but mainly to make the run times consistent and within a reasonable timeout (currently set to 15 minutes). Our main goal is to only ensure we did not break anything, not to actually find bugs.
> 
> We typically choose the fixed seed value as 1 for sanity CI, but the design and DV are in constant flux. We may have to adjust the tests (or adjust the seed even) depending on what the run time looks like.

Understood, I see where the seed coverage comes from now.  Thanks for the explanations @sriyerg and @imphil.

**Comment by sriyerg** [comment]
[Here's](https://dev.azure.com/lowrisc/opentitan-private/_build/results?buildId=9064&view=results) another PR whose private CI failed due to exceeding the 15 minute timeout. 

IMO we should not only fix the seed value for the runs, but also exceed the TO to 20 minutes just to be on the safe side.

**Comment by imphil** [comment]
I just increased the timeout to 20 min.

**Comment by imphil** [comment]
And we now use a fixed see of 1.

**Comment by imphil** [comment]
We now see stable test times of slightly over 7 minutes for the regression runs.

**Comment by sriyerg** [comment]
Great! Thanks.
